rest-api/src/main/scala/za/co/absa/enceladus/rest_api/services/v3/SchemaServiceV3.scala

```diff
@@ -0,0 +1,96 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.enceladus.rest_api.services.v3
+
+import org.apache.spark.sql.types.StructType
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Service
+import za.co.absa.enceladus.model.{Schema, UsedIn, Validation}
+import za.co.absa.enceladus.rest_api.repositories.{DatasetMongoRepository, MappingTableMongoRepository, SchemaMongoRepository}
+import za.co.absa.enceladus.rest_api.services.{SchemaService, VersionedModelService}
+import za.co.absa.enceladus.rest_api.utils.converters.SparkMenasSchemaConvertor
+import scala.concurrent.ExecutionContext.Implicits.global
+
+
+import scala.concurrent.Future
+
+@Service
+class SchemaServiceV3 @Autowired()(schemaMongoRepository: SchemaMongoRepository,
+                                   mappingTableMongoRepository: MappingTableMongoRepository,
+                                   datasetMongoRepository: DatasetMongoRepository,
+                                   sparkMenasConvertor: SparkMenasSchemaConvertor)
+  extends VersionedModelService(schemaMongoRepository) {
+
+  // same as v2
+  override def getUsedIn(name: String, version: Option[Int]): Future[UsedIn] = {
+    for {
+      usedInD <- datasetMongoRepository.findRefEqual("schemaName", "schemaVersion", name, version)
+      usedInM <- mappingTableMongoRepository.findRefEqual("schemaName", "schemaVersion", name, version)
+    } yield UsedIn(Some(usedInD), Some(usedInM))
+  }
+
+  // same as v2
+  def schemaUpload(username: String, schemaName: String, schemaVersion: Int, fields: StructType): Future[(Schema, Validation)] = {
+    super.update(username, schemaName, schemaVersion)({ oldSchema =>
+      oldSchema.copy(fields = sparkMenasConvertor.convertSparkToMenasFields(fields.fields).toList)
+    }).map(_.getOrElse(throw new IllegalArgumentException("Failed to derive new schema from file!")))
+  }
+
+  override def validate(item: Schema): Future[Validation] = {
+    if (item.fields.isEmpty) {
+      // V3 disallows empty schema fields - V2 allowed it at first that to get updated by an attachment upload/remote-load
+      Future.successful(Validation.empty.withError("schema-fields","No fields found! There must be fields defined for actual usage."))
+    } else {
+      Future.successful(Validation.empty)
+    }
+  }
+
+  // same as V2, but fields from payload are used, too
+  override def update(username: String, schema: Schema): Future[Option[(Schema, Validation)]] = {
```

Comment:
It could be that the common fields can be in a trait or all these methods are in the same service but slightly renamed, somethig like: `updateWithFields`

Reply:
I see your point. At this point, I am unsure about the expected lifetime of the V2 in order to strive for easier removal of V2 classes (copy-like) or extracting commons features somewhere.

Reply:
Thinking about this some more, your idea is actually better. I have implemented inspired by your comment. 🙇‍♂️